File: network/session.py
"""Substation session management and state machine."""
from __future__ import annotations
import asyncio
import time
from dataclasses import dataclass, field
from enum import Enum, auto
from typing import Optional
from protocol.frames import ConfigFrame


class SessionState(Enum):
    CONNECTED = auto()         # At least one pipe connected
    CFG1_RECEIVED = auto()     # CFG-1 received from substation
    CFG2_SENT = auto()         # CFG-2 sent to substation
    STREAMING = auto()         # Real-time data flowing
    DISCONNECTED = auto()      # All connections closed


@dataclass
class SubStationSession:
    """Represents one connected substation with its management and data pipes."""
    idcode: str
    version: int = 0                    # Detected from first frame
    peer_ip: str = ""                   # Remote IP for V2 pairing
    state: SessionState = SessionState.CONNECTED

    # Asyncio streams (set when pipes connect)
    mgmt_reader: Optional[asyncio.StreamReader] = field(default=None, repr=False)
    mgmt_writer: Optional[asyncio.StreamWriter] = field(default=None, repr=False)
    data_reader: Optional[asyncio.StreamReader] = field(default=None, repr=False)
    data_writer: Optional[asyncio.StreamWriter] = field(default=None, repr=False)

    cfg1: Optional[ConfigFrame] = None
    cfg2: Optional[ConfigFrame] = None

    last_heartbeat: float = field(default_factory=time.time)
    missed_heartbeats: int = 0

    @property
    def mgmt_connected(self) -> bool:
        return self.mgmt_writer is not None and not self.mgmt_writer.is_closing()

    @property
    def data_connected(self) -> bool:
        return self.data_writer is not None and not self.data_writer.is_closing()

    @property
    def fully_connected(self) -> bool:
        return self.mgmt_connected and self.data_connected

    def close(self):
        """Close all connections."""
        for writer in (self.mgmt_writer, self.data_writer):
            if writer and not writer.is_closing():
                writer.close()
        self.state = SessionState.DISCONNECTED

File: crates/pmusim-app/src/network/session.rs
use pmusim_core::protocol::constants::ProtocolVersion;
use pmusim_core::protocol::frame::ConfigFrame;
use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SessionState {
    Connected,
    Cfg1Received,
    Cfg2Sent,
    Streaming,
    Disconnected,
}

pub struct SubStationSession {
    pub idcode: String,
    pub version: ProtocolVersion,
    pub peer_ip: String,
    pub peer_host: String,
    pub peer_mgmt_port: u16,
    pub state: SessionState,

    pub mgmt_reader: Option<OwnedReadHalf>,
    pub mgmt_writer: Option<OwnedWriteHalf>,
    pub data_reader: Option<OwnedReadHalf>,
    pub data_writer: Option<OwnedWriteHalf>,

    pub cfg1: Option<ConfigFrame>,
    pub cfg2: Option<ConfigFrame>,

    pub last_heartbeat: std::time::Instant,
    pub missed_heartbeats: u32,
}

impl SubStationSession {
    pub fn new(idcode: String, version: ProtocolVersion, peer_ip: String) -> Self {
        Self {
            idcode,
            version,
            peer_ip: peer_ip.clone(),
            peer_host: peer_ip,
            peer_mgmt_port: 0,
            state: SessionState::Connected,
            mgmt_reader: None,
            mgmt_writer: None,
            data_reader: None,
            data_writer: None,
            cfg1: None,
            cfg2: None,
            last_heartbeat: std::time::Instant::now(),
            missed_heartbeats: 0,
        }
    }

    pub fn mgmt_connected(&self) -> bool {
        self.mgmt_writer.is_some()
    }

    pub fn data_connected(&self) -> bool {
        self.data_writer.is_some()
    }

    pub fn close(&mut self) {
        // Dropping OwnedWriteHalf / OwnedReadHalf closes the underlying socket.
        self.mgmt_reader.take();
        self.mgmt_writer.take();
        self.data_reader.take();
        self.data_writer.take();
        self.state = SessionState::Disconnected;
    }
}
